src/hb-kbts.cc

```diff
@@ -0,0 +1,248 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ *
+ * Author(s): Khaled Hosny
+ */
+
+#include "hb.hh"
+
+#if HAVE_KBTS
+
+#include "hb-shaper-impl.hh"
+
+#define KB_TEXT_SHAPE_IMPLEMENTATION
+#define KB_TEXT_SHAPE_STATIC
+#define KB_TEXT_SHAPE_NO_CRT
+#define KBTS_MEMSET hb_memset
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wcast-align"
+#pragma GCC diagnostic ignored "-Wunused-function"
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#include "kb_text_shape.h"
+#pragma GCC diagnostic pop
+
+
+hb_kbts_face_data_t *
+_hb_kbts_shaper_face_data_create (hb_face_t *face)
+{
+  kbts_font *kb_font = nullptr;
+  void *data = nullptr;
+  void *scratch = nullptr;
+  void *memory = nullptr;
+  size_t scratch_size = 0;
+  size_t memory_size = 0;
+
+  hb_blob_t *blob = hb_face_reference_blob (face);
+
+  unsigned int blob_length;
+  const char *blob_data = hb_blob_get_data (blob, &blob_length);
+  if (unlikely (!blob_length))
+  {
+    DEBUG_MSG (KBTS, blob, "Empty blob");
+    goto fail;
+  }
+
+  data = hb_malloc (blob_length);
+  if (unlikely (!data))
+    goto fail;
+
+  kb_font = (kbts_font *) hb_malloc (sizeof(kbts_font));
+  if (unlikely (!kb_font))
+  {
+    hb_free (data);
+    goto fail;
+  }
+
+  // Copy the blob data because kbts_ReadFontData() will modify it.
+  hb_memcpy (data, blob_data, blob_length);
+
+  // The `data` and `memory` pointers will be freed by kbts_FreeFont().
+
+  scratch_size = kbts_ReadFontHeader (kb_font, data, blob_length);
+  scratch = hb_malloc (scratch_size);
+  memory_size = kbts_ReadFontData (kb_font, scratch, scratch_size);
+
+  memory = scratch;
+  if (memory_size > scratch_size)
+    memory = hb_realloc (memory, memory_size);
+
+  kbts_PostReadFontInitialize (kb_font, memory, memory_size);
+
+  if (unlikely (!kbts_FontIsValid(kb_font)))
+  {
+    DEBUG_MSG (KBTS, face, "kbts_font creation failed");
+    _hb_kbts_shaper_face_data_destroy ((hb_kbts_face_data_t *) kb_font);
+    kb_font = nullptr;
+    goto fail;
+  }
+
+fail:
+  hb_blob_destroy (blob);
+
+  return (hb_kbts_face_data_t *) kb_font;
+}
+
+void
+_hb_kbts_shaper_face_data_destroy (hb_kbts_face_data_t *data)
+{
+  kbts_font *font = (kbts_font *) data;
+  hb_free (font->FileBase);
+  hb_free (font->GlyphLookupMatrix);
+}
+
+hb_kbts_font_data_t *
+_hb_kbts_shaper_font_data_create (hb_font_t *font)
+{
+  return (hb_kbts_font_data_t *) HB_SHAPER_DATA_SUCCEEDED;
+}
+
+void
+_hb_kbts_shaper_font_data_destroy (hb_kbts_font_data_t *data)
+{
+}
+
+hb_bool_t
+_hb_kbts_shape (hb_shape_plan_t    *shape_plan,
+		hb_font_t          *font,
+		hb_buffer_t        *buffer,
+		const hb_feature_t *features,
+		unsigned int        num_features)
+{
+  hb_face_t *face = font->face;
+  kbts_font *kb_font = (kbts_font *) (const void *) face->data.kbts;
+  if (unlikely (!kb_font))
```

Comment:
I don't think you need this block. Shaper-selector won't call this function if the data creation failed.